mrmustard/lab_dev/sampler.py

```diff
@@ -0,0 +1,192 @@
+                distribution.
+        """
+
+    def sample(self, state: State, n_samples: int = 1000, seed: int | None = None) -> np.ndarray:
+        r"""
+        Returns an array of samples given a state.
+
+        Args:
+            state: The state to sample.
+            n_samples: The number of samples to generate.
+            seed: An optional seed for random sampling.
+
+        Returns:
+            An array of samples such that the shape is ``(n_samples, n_modes)``.
+        """
+        initial_mode = state.modes[0]
+        initial_samples = self.sample_prob_dist(state[initial_mode], n_samples, seed)
+
+        if len(state.modes) == 1:
+            return initial_samples
+
+        unique_samples, counts = np.unique(initial_samples, return_counts=True)
+        ret = []
+        for unique_sample, counts in zip(unique_samples, counts):
+            meas_op = self.povms[self.meas_outcomes.index(unique_sample)].on([initial_mode])
+            reduced_state = (state >> meas_op.dual).normalize()
+            samples = self.sample(reduced_state, counts)
+            for sample in samples:
+                ret.append(np.append([unique_sample], sample))
+        return np.array(ret)
+
+    def sample_prob_dist(
+        self, state: State, n_samples: int = 1000, seed: int | None = None
+    ) -> np.ndarray:
+        r"""
+        Samples a state by computing the probability distribution.
+
+        Args:
+            state: The state to sample.
+            n_samples: The number of samples to generate.
+            seed: An optional seed for random sampling.
+        """
+        rng = np.random.default_rng(seed) if seed else settings.rng
+        return rng.choice(
+            a=list(product(self.meas_outcomes, repeat=len(state.modes))),
+            p=self.probabilities(state),
+            size=n_samples,
+        )
+
+    def _validate_probs(self, probs: Sequence[float], dx: float, atol: float) -> Sequence[float]:
+        r"""
+        Validates that the given probability distribution sums to `1.0` within some
+        tolerance and returns a renormalized probability distribution to account for
+        small numerical errors.
+
+        Args:
+            probs: The probability distribution to validate.
+            dx: The uniform differential for the probability distribution.
+            atol: The absolute tolerance to validate with.
+        """
+        atol = atol or settings.ATOL
+        probs_dx = probs * dx
+        prob_sum = sum(probs_dx)
+        if not math.allclose(prob_sum, 1, atol):
+            raise ValueError(f"Probabilities sum to {prob_sum} and not 1.0.")
+        return math.real(probs_dx / prob_sum)
+
+
+class PNRSampler(Sampler):
+    r"""
+    A sampler for photon-number resolving (PNR) detectors.
+
+    Args:
+        cutoff: The photon number cutoff.
+    """
+
+    def __init__(self, cutoff: int) -> None:
+        super().__init__(list(range(cutoff)), [Number([0], n) for n in range(cutoff)])
```

Comment:
shouldn't it be `Number([0], n).dual`?

Reply:
In any case, this seems inefficient (albeit correct). There's no need to create the POVM elements and later take inner products with them. Of course it gives the correct results, but since we know the measurement basis is just Fock states can't we bypass this mechanism?